### ISSUE TITLE
`stack sdist` fails with several `hs-source-dirs`

### ISSUE THREAD

**Issue by lierdakil** [commercialhaskell/stack#2432]
Example project: https://github.com/lierdakil/pandoc-crossref

```
$ cabal sdist
...
$ tar ztf dist/pandoc-crossref-0.2.2.1.tar.gz | grep ghc- | grep -v /$
pandoc-crossref-0.2.2.1/ghc-8/Text/Pandoc/CrossRef/Util/Settings/Template.hs
pandoc-crossref-0.2.2.1/ghc-7/Text/Pandoc/CrossRef/Util/Settings/Template.hs
$ stack sdist
...
$ tar ztf .stack-work/dist/x86_64-linux/Cabal-1.22.5.0/pandoc-crossref-0.2.2.1.tar.gz | grep ghc- | grep -v /$
pandoc-crossref-0.2.2.1/ghc-8/Text/Pandoc/CrossRef/Util/Settings/Template.hs
pandoc-crossref-0.2.2.1/ghc-8/Text/Pandoc/CrossRef/Util/Settings/Template.hs
```

Note that with `stack sdist` , `ghc-7` version is omitted, and there are two copies of `ghc-8` version instead.

Ordering is important.

EDIT:

I was under impression that this is caused by cabal conditionals. It is not. With three `hs-source-dirs` last one is replaced with second-to-last.

**Comment by lierdakil** [comment]
I was able to track the problem down to this invocation:
```
$HOME/.stack/setup-exe-cache/x86_64-linux/setup-Simple-Cabal-1.22.5.0-ghc-7.10.3 --builddir=<...> sdist --list-sources /tmp/<...>/source-files-list
```

So this might be a bug in cabal-1.22.5.0.

At the very least I can confirm that this doesn't happen with `nightly-2016-07-14` resolver (which uses Cabal-1.24.0.0 I believe).

After some further testing, it appears that using `stack setup --upgrade-cabal` also helps.

Question: why not automatically use cabal-1.22.8.0 in lts resolver?

**Comment by mgsloan** [comment]
It is tricky to say whether we want to be able to upgrade Cabal or use a consistent one from the snapshot.  Currently it is upgradeable.

Interesting to know that this is potentially a Cabal issue.  Perhaps stack should have a Cabal version blacklist.

**Comment by lierdakil** [comment]
I think I've found relevant cabal issues: https://github.com/haskell/cabal/issues/367, https://github.com/haskell/cabal/issues/2544.

As far as I can tell, this was fixed only in 1.24; the issue seems to be present in Cabal-1.22.8 at least:

```
$ cabal --version
cabal-install version 1.22.9.0
using version 1.22.8.0 of the Cabal library 
$ cabal sdist
Source tarball created: dist/pandoc-crossref-0.2.3.0.tar.gz
$ tar ztf dist/pandoc-crossref-0.2.3.0.tar.gz | grep ghc- | grep -v /$
pandoc-crossref-0.2.3.0/ghc-8/Text/Pandoc/CrossRef/Util/Settings/Template.hs
```

Notice though that `stack` includes a directory twice (compare with OP), while `cabal` does that only once, so maybe `nub`ing file list before packing might be a good idea.

Not sure if you would consider this issue valid at this point, since it's quite obviously not stack's fault, and auto-upgrading Cabal to another major version doesn't sound like a particularly good idea.
